## [mlcourse.ai](https://mlcourse.ai) – Open Machine Learning Course 

Author: Mariya Mansurova, Analyst & developer in Yandex.Metrics team. Translated by Ivan Zakharov, ML enthusiast. <br>This material is subject to the terms and conditions of the [Creative Commons CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/) license. Free use is permitted for any non-commercial purpose.

# <center> Assignment #9 (demo)
## <center> Time series analysis

**Fill cells marked with "Your code here" and submit your answers to the questions through the [web form](https://docs.google.com/forms/d/1UYQ_WYSpsV3VSlZAzhSN_YXmyjV7YlTP8EYMg8M8SoM/edit).**

In [1]:
import os
import warnings
warnings.filterwarnings('ignore')
import numpy as np
import pandas as pd

from plotly import __version__
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot
from plotly import graph_objs as go
import requests

print(__version__) # need 1.9.0 or greater
init_notebook_mode(connected = True)


def plotly_df(df, title = ''):
    data = []
    
    for column in df.columns:
        trace = go.Scatter(
            x = df.index,
            y = df[column],
            mode = 'lines',
            name = column
        )
        data.append(trace)
    
    layout = dict(title = title)
    fig = dict(data = data, layout = layout)
    iplot(fig, show_link=False)

3.6.1


## Data preparation

In [2]:
df = pd.read_csv('../data/wiki_machine_learning.csv', sep = ' ')
df = df[df['count'] != 0]
df.head()

date  count lang              page  rank   month             title
81  2015-01-01   1414   en  Machine_learning  8708  201501  Machine_learning
80  2015-01-02   1920   en  Machine_learning  8708  201501  Machine_learning
79  2015-01-03   1338   en  Machine_learning  8708  201501  Machine_learning
78  2015-01-04   1404   en  Machine_learning  8708  201501  Machine_learning
77  2015-01-05   2264   en  Machine_learning  8708  201501  Machine_learning

In [3]:
df.shape

(383, 7)

## Predicting with FB Prophet
We will train at first 5 months and predict the number of trips for June.

In [4]:
df.date = pd.to_datetime(df.date)

In [5]:
plotly_df(df.set_index('date')[['count']])

In [6]:
from fbprophet import Prophet

In [7]:
predictions = 30

df = df[['date', 'count']]
df.columns = ['ds', 'y']
df.tail()

ds     y
382 2016-01-16  1644
381 2016-01-17  1836
376 2016-01-18  2983
375 2016-01-19  3389
372 2016-01-20  3559

**<font color='red'>Question 1:</font>** What is the prediction of the number of views of the wiki page on January 20? Round to the nearest integer.

- 4947
- 3426
- 5229
- 2744

In [8]:
# You code here
train_df = df[:-predictions]
model = Prophet()
model.fit(train_df);

INFO:fbprophet:Disabling yearly seasonality. Run prophet with yearly_seasonality=True to override this.
INFO:fbprophet:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


In [9]:
future = model.make_future_dataframe(periods=predictions)
preds = model.predict(future)

In [10]:
jan_20_2016 = '2016-01-20'
print(f'Prediction for {jan_20_2016} is {round(preds[preds["ds"]==jan_20_2016]["yhat"].values[0])}')
# preds

Prediction for 2016-01-20 is 3426.0


Estimate the quality of the prediction with the last 30 points.

In [11]:
def make_comparison_dataframe(historical, forecast):
    """Join the history with the forecast.
    
       The resulting dataset will contain columns 'yhat', 'yhat_lower', 'yhat_upper' and 'y'.
    """
    return forecast.set_index('ds')[['yhat', 'yhat_lower', 'yhat_upper']].join(historical.set_index('ds'))

In [12]:
def calculate_forecast_errors(df, prediction_size):
    """Calculate MAPE and MAE of the forecast.
    
       Args:
           df: joined dataset with 'y' and 'yhat' columns.
           prediction_size: number of days at the end to predict.
    """
    
    # Make a copy
    df = df.copy()
    
    # Now we calculate the values of e_i and p_i according to the formulas given in the article above.
    df['e'] = df['y'] - df['yhat']
    df['p'] = 100 * df['e'] / df['y']
    
    # Recall that we held out the values of the last `prediction_size` days
    # in order to predict them and measure the quality of the model. 
    
    # Now cut out the part of the data which we made our prediction for.
    predicted_part = df[-prediction_size:]
    
    # Define the function that averages absolute error values over the predicted part.
    error_mean = lambda error_name: np.mean(np.abs(predicted_part[error_name]))
    
    # Now we can calculate MAPE and MAE and return the resulting dictionary of errors.
    return {'MAPE': error_mean('p'), 'MAE': error_mean('e')}

In [13]:
# You code here
cmp_df = make_comparison_dataframe(df, preds)
cmp_df.tail(n=3)

for err_name, err_value in calculate_forecast_errors(cmp_df, predictions).items():
    print(f'{err_name}: {round(err_value,2)}')

MAPE: 34.5
MAE: 599.84


**<font color='red'>Question 2:</font> What is MAPE equal to?**

- 34.5
- 42.42
- 5.39
- 65.91

**<font color='red'>Question 3:</font> What is MAE equal to?**

- 355
- 4007
- 600
- 903

## Predicting with ARIMA

In [14]:
%matplotlib inline
import matplotlib.pyplot as plt
from scipy import stats
import statsmodels.api as sm
plt.rcParams['figure.figsize'] = (15, 10)

**<font color='red'>Question 4:</font> Let's verify the stationarity of the series using the Dickey-Fuller test. Is the series stationary? What is the p-value?**

- Series is stationary, p_value = 0.107
- Series is not stationary, p_value = 0.107
- Series is stationary, p_value = 0.001
- Series is not stationary, p_value = 0.001

In [15]:
# You code here
from statsmodels import api as sm

In [16]:
dickey_fuller = sm.tsa.stattools.adfuller(train_df['y'].values)
print(f'Dickey-Fuller p-value: {round(dickey_fuller[1], 3)}')

Dickey-Fuller p-value: 0.107


**Next, we turn to the construction of the SARIMAX model (`sm.tsa.statespace.SARIMAX`).<br> <font color='red'>Question 5:</font> What parameters are the best for the model according to the `AIC` criterion?**

- D = 1, d = 0, Q = 0, q = 2, P = 3, p = 1
- D = 2, d = 1, Q = 1, q = 2, P = 3, p = 1
- D = 1, d = 1, Q = 1, q = 2, P = 3, p = 1
- D = 0, d = 0, Q = 0, q = 2, P = 3, p = 1

In [17]:
parameters_list = [
    {'D': 1, 'd': 0, 'Q': 0},
    {'D': 2, 'd': 1, 'Q': 1},
    {'D': 1, 'd': 1, 'Q': 1},
    {'D': 0, 'd': 0, 'Q': 0}
]
s = 7
q = 2
P = 3
p = 1
best_aic = None

In [18]:
%%time
for params in parameters_list:
    try:
        model = sm.tsa.statespace.SARIMAX(train_df['y'].values, order=(p, params['d'], q), 
                                          seasonal_order=(P, params['D'], params['Q'], s)).fit(disp=-1)
    except Exception as e:
        print(f'Exception occured! {e}')
        continue
    aic = model.aic
    
    if best_aic is None or aic < best_aic:
        best_aic = aic
        best_params = params

/home/artemii/anaconda3/lib/python3.7/site-packages/statsmodels/base/model.py:508: ConvergenceWarning:

Maximum Likelihood optimization failed to converge. Check mle_retvals



Exception occured! Non-stationary starting autoregressive parameters found with `enforce_stationarity` set to True.
CPU times: user 2min 49s, sys: 186 ms, total: 2min 50s
Wall time: 1min 37s


In [19]:
best_params.update(dict(q=q, P=P, p=p))
print(f'Best params: {best_params}\n Best AIC: {best_aic}')

Best params: {'D': 2, 'd': 1, 'Q': 1, 'q': 2, 'P': 3, 'p': 1}
 Best AIC: 4988.954046911636
